pkgs/dart_model/lib/src/scopes.dart

```diff
@@ -0,0 +1,121 @@
+// Copyright (c) 2024, the Dart project authors. Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'dart:async';
+import 'dart:typed_data';
+
+import 'json_buffer/json_buffer_builder.dart';
+
+/// Scope for accumulating `dart_model` data for efficient transmission.
+///
+/// A scope introduces buffers that can be written to. Some `dart_model` types
+/// are then instantiated directly into the buffer instead of on the heap,
+/// making them ready to send with no further copying.
+///
+/// Run code in a scope by calling [run] or [runAsync] on the scope.
+enum Scope {
+  /// No scope.
+  ///
+  /// If code is running in a scope it can use `Scope.none` to run code
+  /// without a scope. For example macro scope is about accumulating
+  /// augmentations, but macros also send queries. The `none` scope must be
+  /// used when sending queries.
+  none,
+
+  /// Server-side scope for handling queries.
+  ///
+  /// Provides a new buffer for the query result.
+  query,
+
+  /// Client-side scope for macro code to run in.
+  ///
+  /// Provides a new buffer for the macro's augmentations.
+  macro,
+
+  /// Temporary scope for type evaluation.
+  ///
+  /// A new buffer is introduced for temporary intermediate values. It's an
+  /// error to try to serialize it.
+  evaluating;
+
+  /// Runs [function] in this scope.
+  T run<T>(T Function() function) {
+    return runZoned(function, zoneValues: {
+      _symbol: _ScopeData._(this, this == none ? null : JsonBufferBuilder())
+    });
+  }
+
+  /// Runs [function] in this scope.
+  Future<T> runAsync<T>(Future<T> Function() function) async {
```

Comment:
Could just return the call to `runZoned` directly here, no need for the async/await afaik?